### PR TITLE
# 25 feature-search product page

### PR DESCRIPTION
**Pull request by Muhet**
#### What does this PR do?

search product

#### Description of Task to be completed?

A buyer needs to have a way to search for a product
t. by name, by price range, by category, etc it can also be a combination of two or more with AND operation.

As a user, I should be able to use the search component
So that, I can easily retrieve records from both the request and approval table

#### How should this be manually tested?
```
git clone https://github.com/atlp-rwanda/ecommerce-app-predators-fe.git
cd ecommerce-app-predators-fe
git checkout branch-?
npm i
npm run dev
```

#### What are the relevant pivotal tracker/Trello stories?
Trello [Link]() (#-25)

#### Screenshots (if appropriate)
####ALL PRODUCTS
|
search product ONe
|
#### SEARCHED PRODUCT
|
search product Two

|

### CODE REVIEW

**Review by ALVINdimpos — APPROVED** [review]

File: src/components/cardComponent/Card.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import { useEffect } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import { fetchProducts } from '../../redux/action/ProductAction';
+
+interface Product {
+  id: any;
```
Comment (chanceDira) [inline src/components/cardComponent/Card.tsx]: Use the interfaces

File: src/components/cardComponent/Card.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  name: any;
+  price: any;
+  picture_urls: any;
+  rating: any;
+}
+
+export const Card = (product: Product) => {
+  const products = useSelector(
+    (state: { products: { data: any } }) => state.products.data
```
Comment (chanceDira) [inline src/components/cardComponent/Card.tsx]: Use the interfaces

File: src/components/cardComponent/Card.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  );
+
+  console.log(products);
+  const error = useSelector(
+    (state: { products: { error: any } }) => state.products.error
```
Comment (chanceDira) [inline src/components/cardComponent/Card.tsx]: Use the interfaces

File: src/pages/SearchProductPage.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import { useDispatch, useSelector } from 'react-redux';
+import { Dispatch } from 'redux';
+import { fetchProducts } from '../redux/action/ProductAction';
+import { searchProducts } from '../redux/action/SearchAction';
+import Sidebar from '../components/userSidebar/Sidebar';
+import { Card } from './../components/cardComponent/Card';
+import React from 'react';
+
+interface Product {
+  id: any;
```
Comment (chanceDira) [inline src/pages/SearchProductPage.tsx]: Use the interfaces

File: src/pages/SearchProductPage.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  name: any;
+  price: any;
+  picture_urls: any;
+}
+
+function SearchProductPage() {
+  const products = useSelector(
+    (state: { products: { data: any } }) => state.products.data
```
Comment (chanceDira) [inline src/pages/SearchProductPage.tsx]: Use the interface

File: src/redux/action/ProductAction.tsx [hunk 6]
```diff
@@ -18,36 +17,35 @@ type ProductData = {
...
     }
   }
 );
 
-
-
 export const addProduct = createAsyncThunk(
-  "products/addProduct",
+  'products/addProduct',
   async (product: any, thunkAPI) => {
```
Comment (chanceDira) [inline src/redux/action/ProductAction.tsx:49]: Use the interface

**Review by chanceDira — CHANGES_REQUESTED** [review]
Good job @Muhet 
I left some comments below

**Review by vanessa-abewe — APPROVED** [review]

**Review by NTElissa — APPROVED** [review]
GOOD PROGRESS

**Review by yobuir — APPROVED** [review]

File: src/components/cardComponent/Card.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
+import { fetchProducts } from '../../redux/action/ProductAction';
+
+interface Product {
+  id: number;
+  name: string;
+  price: string;
+  picture_urls: string;
+  rating: any;
```
Comment (chanceDira) [inline src/components/cardComponent/Card.tsx:10]: rating is having any?

File: src/redux/action/SearchAction.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      }
+
+      const response = await axios.get(url);
+      return response.data;
+    } catch (error) {
+      throw new Error('Error searching for products');
+    }
+  }
+);
+/* export const searchProducts = createAsyncThunk(
```
Comment (chanceDira) [inline src/redux/action/SearchAction.tsx]: Can you remove these codes, if we are not using it

**Review by chanceDira — CHANGES_REQUESTED** [review]
Below are finals remarks to fix

File: src/components/cardComponent/Card.tsx (hunk 7 above)
Comment (Muhet) [reply src/components/cardComponent/Card.tsx:10]: It is not declared in our database ( We don't have it their)

**Review by chanceDira — APPROVED** [review]
Looking good

**Pull request merged**